### PR TITLE
Added CREATE INDEX

### PR DESCRIPTION
**Pull request by darionyaphet**

### CODE REVIEW

File: src/planner/Maintain.h [hunk 1]
```diff
@@ -341,71 +341,180 @@ class DropSchema : public SingleInputNode {
...
+
+//     std::unique_ptr<cpp2::PlanNodeDescription> explain() const override;
+
+// protected:
+//     std::string            name_;
+//     meta::cpp2::Schema     schema_;
+//     bool                   ifNotExists_;
+// };
+
+class CreateTagIndexNode final : public SingleInputNode {
```
Comment (dutor) [inline src/planner/Maintain.h:416]: Why remove the `CreateIndex` class? I think it's more natural to reuse some code that way.

Comment (darionyaphet) [reply src/planner/Maintain.h:416]: Yep Actually this one is still processing and I push it out just wish to avoid some one implement it duplicated.

**Comment by bright-starry-sky** [comment]
Ready to review ?

File: src/validator/MaintainValidator.cpp [hunk 2]
```diff
@@ -51,7 +51,7 @@ Status CreateTagValidator::validateImpl() {
 
 Status CreateTagValidator::toPlan() {
     auto *plan = qctx_->plan();
-    auto doNode = CreateTag::make(qctx_,
+    auto doNode = CreateTagNode::make(qctx_,
```
Comment (bright-starry-sky) [inline src/validator/MaintainValidator.cpp:54]: Is `genSingleNodePlan` better？

**Review by bright-starry-sky — COMMENTED** [review]
I think , we can unify all tagIndexXXX and edgeIndexXXX method to indexXXX. It makes the code a little bit cleaner. Good for later maintenance. WDYT?

File: src/executor/maintain/TagIndexExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    return Status::OK();
+}
+
+folly::Future<Status> DescTagIndexExecutor::execute() {
+    return Status::OK();
+}
+
+folly::Future<Status> RebuildTagIndexExecutor::execute() {
+    return Status::OK();
+}
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/executor/maintain/TagIndexExecutor.cpp:27]: Still Processing thanks~

File: src/executor/maintain/EdgeExecutor.cpp [hunk 4]
```diff
@@ -16,12 +16,12 @@ namespace graph {
 folly::Future<Status> CreateEdgeExecutor::execute() {
     SCOPED_TIMER(&execTime_);
 
-    auto *ceNode = asNode<CreateEdge>(node());
+    auto *ceNode = asNode<CreateEdgeNode>(node());
```
[the remark this answers is not in the document]
Comment (CPWstatic) [reply src/executor/maintain/EdgeExecutor.cpp:19]: Agree, we don't need the `Node` suffix.

**Review by dutor — COMMENTED** [review]
Thanks for the big efforts!

Could you please focus on the index creation first? So that we can make progress on other index-related modules.

File: src/executor/maintain/EdgeExecutor.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dutor) [reply src/executor/maintain/EdgeExecutor.cpp:19]: Agree

[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/executor/maintain/EdgeExecutor.cpp:19]: I found `CreateSchemaNode` so append Node as suffix. Is this a mistake?

File: src/executor/maintain/TagIndexExecutor.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    CreateTagIndexExecutor(const PlanNode *node, QueryContext *ectx)
+        : Executor("CreateTagIndexExecutor", node, ectx) {}
+
+    folly::Future<Status> execute() override;
+};
+
+class DropTagIndexExecutor final : public Executor {
+public:
+    DropTagIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("DropTagIndexExecutor", node, ectx) {}
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.h:26]: indentation

File: src/executor/maintain/TagIndexExecutor.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    DescTagIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("DescTagIndexExecutor", node, ectx) {}
+
+    folly::Future<Status> execute() override;
+};
+
+class RebuildTagIndexExecutor final : public Executor {
+public:
+    RebuildTagIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("RebuildTagIndexExecutor", node, ectx) {}
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.h:42]: ditto

Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.h:42]: ditto

File: src/executor/maintain/TagIndexExecutor.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    DropTagIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("DropTagIndexExecutor", node, ectx) {}
+
+    folly::Future<Status> execute() override;
+};
+
+class DescTagIndexExecutor final : public Executor {
+public:
+    DescTagIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("DescTagIndexExecutor", node, ectx) {}
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.h:34]: ditto

File: src/validator/MaintainValidator.cpp (hunk 2 above)
Comment (darionyaphet) [reply src/validator/MaintainValidator.cpp]: maybe they are the same or I will unify them later

File: src/planner/Maintain.cpp [hunk 8]
```diff
@@ -31,18 +31,46 @@ std::unique_ptr<cpp2::PlanNodeDescription> AlterSchemaNode::explain() const {
...
     addDescription("ifExists", util::toJson(ifExists_), desc.get());
     return desc;
 }
 
+std::unique_ptr<cpp2::PlanNodeDescription> CreateIndexNode::explain() const {
+    auto desc = SingleInputNode::explain();
+    addDescription("schemaName", schemaName_, desc.get());
+    addDescription("indexName", indexName_, desc.get());
+    addDescription("fields", folly::toJson(util::toJson(fields_)), desc.get());
+    addDescription("ifNotExists", folly::to<std::string>(ifNotExists_), desc.get());
```
Comment (laura-ding) [inline src/planner/Maintain.cpp:52]: util::toJson(ifNotExists_)

File: src/executor/maintain/EdgeIndexExecutor.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    CreateEdgeIndexExecutor(const PlanNode *node, QueryContext *ectx)
+        : Executor("CreateEdgeIndexExecutor", node, ectx) {}
+
+    folly::Future<Status> execute() override;
+};
+
+class DropEdgeIndexExecutor final : public Executor {
+public:
+    DropEdgeIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("DropEdgeIndexExecutor", node, ectx) {}
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.h:26]: indentation

File: src/executor/maintain/EdgeIndexExecutor.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    folly::Future<Status> execute() override;
+};
+
+class DescEdgeIndexExecutor final : public Executor {
+public:
+    DescEdgeIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("DescEdgeIndexExecutor", node, ectx) {}
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.h:34]: ditto

File: src/executor/maintain/EdgeIndexExecutor.h [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    folly::Future<Status> execute() override;
+};
+
+class RebuildEdgeIndexExecutor final : public Executor {
+public:
+    RebuildEdgeIndexExecutor(const PlanNode *node, QueryContext *ectx)
+            : Executor("RebuildEdgeIndexExecutor", node, ectx) {}
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.h:42]: ditto

File: src/validator/MaintainValidator.cpp [hunk 12]
```diff
@@ -51,8 +52,8 @@ Status CreateTagValidator::validateImpl() {
 
 Status CreateTagValidator::toPlan() {
     auto *plan = qctx_->plan();
-    auto doNode = CreateTag::make(qctx_,
-            plan->root(), std::move(name_), std::move(schema_), ifNotExist_);
+    auto doNode = CreateTag::make(qctx_, plan->root(), std::move(name_),
+                                      std::move(schema_), ifNotExist_);
```
Comment (laura-ding) [inline src/validator/MaintainValidator.cpp:56]: indentation

File: src/planner/Maintain.h [hunk 13]
```diff
@@ -333,80 +331,317 @@ class DropSchema : public SingleInputNode {
...
 };
 
-class BuildTagIndex final : public SingleInputNode {
+class RebuildTagIndex final : public RebuildIndexNode {
 public:
+    static RebuildTagIndex* make(QueryContext* qctx,
+                                 PlanNode* input,
+                                 std::string indexName) {
+        return qctx->objPool()->add(new RebuildTagIndex(qctx->genId(), input,
+                                                            std::move(indexName)));
```
Comment (laura-ding) [inline src/planner/Maintain.h:621]: indentation

File: src/planner/Maintain.cpp (hunk 8 above)
Comment (darionyaphet) [reply src/planner/Maintain.cpp:52]: ifNotExists_ is bool ?

File: src/validator/MaintainValidator.cpp (hunk 12 above)
Comment (darionyaphet) [reply src/validator/MaintainValidator.cpp:56]: Good Cache~

**Review by dutor — APPROVED** [review]
LGTM.

**Review by bright-starry-sky — APPROVED** [review]
LGTM.

**Pull request merged**